File: jay/nodejs/producer.js
const { Kafka } = require("kafkajs");
const os = require("os");
require("dotenv").config(); // Load environment variables

// Kafka setup
const kafka = new Kafka({
  clientId: os.hostname(),
  brokers: process.env.BROKERS.split(","), // Brokers from .env
});

// Configuration from .env with debug logs
console.log("Loading configuration from .env...");
const topicName = process.env.TOPIC_NAME || "default-topic";
const numThreads = parseInt(process.env.NUM_THREADS, 10) || 1;
const numMessages = parseInt(process.env.NUM_MESSAGES, 10) || 100;
const kpiThreshold = parseInt(process.env.KPI_THRESHOLD, 10) || 50;
const messageInterval = parseInt(process.env.MESSAGE_INTERVAL, 10) * 1000 || 1000;

console.log("Configuration:");
console.log(`  Topic Name: ${topicName}`);
console.log(`  Number of Threads: ${numThreads}`);
console.log(`  Messages per Thread: ${numMessages}`);
console.log(`  Message Interval: ${messageInterval}ms`);

const producer = kafka.producer();

// Helper function for delay
const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

// Message production logic
const produceMessages = async () => {
  await producer.connect();
  console.log("Producer connected");

  for (let threadId = 0; threadId < numThreads; threadId++) {
    console.log(`Starting thread ${threadId}`);
    for (let i = 0; i < numMessages; i++) {
      const producerTimestamp = new Date().toISOString();

      const message = {
        uuid: `${threadId}-${i}-${Date.now()}`,
        data: `sample_data_from_thread_${threadId}_${i}`,
        producer_instance: os.hostname(),
        producer_timestamp: producerTimestamp,
      };

      // Send message to Kafka
      await producer.send({
        topic: topicName,
        messages: [
          {
            key: message.uuid,
            value: JSON.stringify(message),
          },
        ],
      });

      // Debug log after sending each message
      console.log(`Thread ${threadId}: Sent message ${message.uuid} at ${producerTimestamp}`);

      // Simulate latency calculation and KPI check
      const consumerTimestamp = new Date(); // Assume instant processing for simulation
      const latency = new Date(consumerTimestamp) - new Date(producerTimestamp);

      const meetsKpi = latency <= kpiThreshold;

      // Show result in terminal
      if (meetsKpi) {
        console.log(`✅ Message ${message.uuid} meets KPI: Latency = ${latency}ms`);
      } else {
        console.log(`❌ Message ${message.uuid} exceeds KPI: Latency = ${latency}ms`);
      }

      // Debug log before sleeping
      console.log(`Thread ${threadId}: Sleeping for ${messageInterval}ms before next message`);
      await sleep(messageInterval);
    }
    console.log(`Thread ${threadId} finished`);
  }

  await producer.disconnect();
  console.log("Producer finished");
};

// Run the producer
produceMessages().catch((err) => {
  console.error("Error in producer:", err);
});

File: jay/python/producer.py
from confluent_kafka import Producer
import threading
import time
import uuid
import json
import socket  # To get the hostname of the EC2 instance
from datetime import datetime
from dotenv import load_dotenv  # Import dotenv to load environment variables
import os

# Load environment variables from .env file
load_dotenv()

# Kafka producer configuration
conf = {
    'bootstrap.servers': os.getenv('BOOTSTRAP_SERVERS'),
    'linger.ms': int(os.getenv('LINGER_MS')),
    'batch.size': int(os.getenv('BATCH_SIZE')),
    'compression.type': os.getenv('COMPRESSION_TYPE'),
    'acks': os.getenv('ACKS'),
}

# Initialize Kafka producer
producer = Producer(conf)

# Get the EC2 instance name
producer_instance_name = socket.gethostname()

# Producer function that runs in each thread
def produce_messages(thread_id, num_messages, message_interval):
    for _ in range(num_messages):
        message = {
            "uuid": str(uuid.uuid4()),
            "data": f"sample_data_from_thread_{thread_id}",
            "producer_instance": producer_instance_name,
            "producer_timestamp": datetime.utcnow().isoformat()
        }
        # Convert message to JSON string
        producer.produce(os.getenv('TOPIC_NAME'), key=message["uuid"], value=json.dumps(message))
        # Call poll to handle delivery callbacks
        producer.poll(0)
        time.sleep(message_interval)  # Control message rate

# Function to periodically call poll to clear out the queue
def poll_producer():
    while True:
        producer.poll(float(os.getenv('POLL_INTERVAL')))  # Processes any delivery callbacks in the background

# Start the poll thread
poll_thread = threading.Thread(target=poll_producer, daemon=True)
poll_thread.start()

# Number of threads, messages per thread, and message interval
num_threads = int(os.getenv('NUM_THREADS'))
num_messages = int(os.getenv('NUM_MESSAGES'))
message_interval = float(os.getenv('MESSAGE_INTERVAL'))

# Start producer threads
threads = []
for i in range(num_threads):
    thread = threading.Thread(target=produce_messages, args=(i, num_messages, message_interval))
    threads.append(thread)
    thread.start()

# Wait for all threads to finish
for thread in threads:
    thread.join()

# Flush any remaining messages
producer.flush()
